chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,467 @@
+            - Calculate variance term
+            - Calculate distance term
+            - Calculate regularization term
+            - Add weights to all and return loss value
+
+        Args:
+            x (tuple) : Contains several inputs
+                - x[0] (:class:`~chainer.Variable` or \
+                        :class:`numpy.ndarray` or \
+                        :class:`cupy.ndarray`)
+                        segmentation prediction output
+                - x[1] (:class:`~chainer.Variable` or \
+                        :class:`numpy.ndarray` or \
+                        :class:`cupy.ndarray`)
+                        segmentation ground truth
+                - x[2] (:class:`~chainer.Variable` or \
+                        :class:`numpy.ndarray` or \
+                        :class:`cupy.ndarray`)
+                        number of objects in ground truth
+                - x[3] (:class:`~chainer.Variable` or \
+                        :class:`numpy.ndarray` or \
+                        :class:`cupy.ndarray`)
+                        indexes of non-zero ground truths
+        Returns:
+            (:class:`~chainer.Variable` or \
+            :class:`numpy.ndarray` or \
+            :class:`cupy.ndarray`) : Loss value
+        """
+
+        buffer = chainer.config.type_check
+        chainer.config.type_check = False
+
+        assert(len(args) == 4)
+
+        # Inputs
+        prediction, labels = self._prepare_inputs(args[0], args[1])
+        n_objects, gt_idx = args[2:4]
+
+        # Calculate cluster means
+        c_means = self._means(prediction, labels, n_objects,
+                              self.max_n_clusters, gt_idx)
+
+        # Calculate losses
+        l_var = self._variance_term(prediction, labels, c_means, self.delta_v,
+                                    gt_idx)
+        l_dist = self._distance_term(c_means, self.delta_d, n_objects)
+        l_reg = self._regularization_term(c_means, gt_idx)
+
+        chainer.config.type_check = buffer
+        return self.alpha * l_var + self.beta * l_dist + self.gamma * l_reg
+
+
+def discriminative_margin_based_clustering_loss(
+        x, delta_v, delta_d, max_n_clusters,
+        norm=1, alpha=1.0, beta=1.0, gamma=0.001):
+    """Discriminative margin based clustering loss function
+
+    This is the implementation of following paper:
+    https://arxiv.org/pdf/1802.05591.pdf
+
+    In segmentation, one of the biggest problem is having noise at the output
+    of a trained network.
+    For cross-entropy based approaches, if the pixel value is wrong,
+    the loss value will be same independent from wrong pixel's location.
+    However, for segmentation, even though network gives wrong pixel output,
+    it is desirable to have it
+    as close as possible to the original position.
+    By applying discriminative loss function, groups of segmentation instances
+    can be moved together.
+
+    This loss function calculates three different parameters:
```

Comment:
<img width="814" alt="2018-06-20 19 31 00" src="https://user-images.githubusercontent.com/407052/41653841-4ba83f38-74c2-11e8-9de3-168dab46f15c.png">

Reply:
Could you remove the empty line of l.415?